File: app/src/main/java/com/ocram/qichwadic/data/datasource/SearchLocalDataStore.java
package com.ocram.qichwadic.data.datasource;

import androidx.sqlite.db.SimpleSQLiteQuery;

import com.ocram.qichwadic.data.repository.SearchRepositoryImpl;
import com.ocram.qichwadic.domain.interactor.SearchType;
import com.ocram.qichwadic.domain.model.Definition;
import com.ocram.qichwadic.domain.model.SearchResult;
import com.ocram.qichwadic.framework.dao.SearchDao;
import com.ocram.qichwadic.presentation.viewmodel.viewstate.SearchParams;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Flowable;

public class SearchLocalDataStore implements SearchRepositoryImpl.SearchLocalDataStore {

    private final SearchDao searchDao;

    @Inject
    public SearchLocalDataStore(SearchDao searchDao) {
        this.searchDao = searchDao;
    }

    @Override
    public Flowable<List<SearchResult>> getDictionariesContainingWord(String langBegin, String landEnd, String wordQuery) {
        return searchDao.getDictionariesContainingWord(langBegin, landEnd, wordQuery);
    }

    @Override
    public Flowable<List<Definition>> findDefinitionsInDictionary(List<Integer> dictionaryIds, String wordQuery) {
        return searchDao.findDefinitionsInAllDictionaries(buildCompatibleWritingWord(dictionaryIds, wordQuery));
    }

    private SimpleSQLiteQuery buildCompatibleWritingWord(List<Integer> dictionaryIds, String word) {
        String query =
                "SELECT * FROM " +
                        "(SELECT de.* " +
                        "FROM definition de INNER JOIN dictionary di ON de.dictionary_id = di.id " +
                        "WHERE di.id = %s AND de.word LIKE \"%s\" " +
                        "ORDER BY de.id LIMIT 20)";
        StringBuilder queryBuilder = new StringBuilder();
        for(int i = 0; i < dictionaryIds.size(); i++) {
            queryBuilder.append(String.format(query, dictionaryIds.get(i), word));
            if(i< dictionaryIds.size() -1) {
                queryBuilder.append(" UNION ALL ");
            }
        }
        return new SimpleSQLiteQuery(queryBuilder.toString());
    }

    @Override
    public Flowable<List<Definition>> fetchMoreResults(int dictionaryId, String word, int searchType, int page) {
        return searchDao.findDefinitionsLike(dictionaryId, SearchType.buildSearchCriteria(searchType, word), computeOffset(page));
    }

    private int computeOffset(int page){
        return SearchParams.MAX_SEARCH_RESULTS * (page - 1 );
    }
}

File: app/src/main/java/com/ocram/qichwadic/features/search/data/datastore/SearchLocalDataStore.kt
package com.ocram.qichwadic.features.search.data.datastore

import androidx.sqlite.db.SimpleSQLiteQuery

import com.ocram.qichwadic.core.data.model.DefinitionEntity
import com.ocram.qichwadic.core.data.local.dao.SearchDao
import com.ocram.qichwadic.core.data.model.SearchResultEntity
import com.ocram.qichwadic.features.search.data.SearchType

interface SearchLocalDataStore {

    suspend fun getDictionariesContainingWord(langBegin: String, landEnd: String, wordQuery: String): List<SearchResultEntity>

    suspend fun findDefinitionsInDictionary(dictionaryIds: Collection<Int>, wordQuery: String): List<DefinitionEntity>

    suspend fun fetchMoreResults(dictionaryId: Int, word: String, searchType: Int, page: Int): List<DefinitionEntity>
}

class SearchLocalDataStoreImpl(private val searchDao: SearchDao): SearchLocalDataStore {

    private val maxSearchResults = 20

    override suspend fun getDictionariesContainingWord(langBegin: String, landEnd: String, wordQuery: String): List<SearchResultEntity> {
        return searchDao.getDictionariesContainingWord(langBegin, landEnd, wordQuery)
    }

    override suspend fun findDefinitionsInDictionary(dictionaryIds: Collection<Int>, wordQuery: String): List<DefinitionEntity> {
        return searchDao.findDefinitionsInAllDictionaries(buildCompatibleWritingWord(dictionaryIds, wordQuery))
    }

    private fun buildCompatibleWritingWord(dictionaryIds: Collection<Int>, word: String): SimpleSQLiteQuery {
        val query = "SELECT * FROM " +
                "(SELECT de.* " +
                "FROM definition de INNER JOIN dictionary di ON de.dictionary_id = di.id " +
                "WHERE di.id = %s AND de.word LIKE \"%s\" " +
                "ORDER BY de.id LIMIT 20)"
        val queryBuilder = StringBuilder()
        dictionaryIds.forEachIndexed { index, dictionaryId ->
            queryBuilder.append(query.format(dictionaryId, word))
            if (index < dictionaryIds.size - 1) {
                queryBuilder.append(" UNION ALL ")
            }
        }
        return SimpleSQLiteQuery(queryBuilder.toString())
    }

    override suspend fun fetchMoreResults(dictionaryId: Int, word: String, searchType: Int, page: Int): List<DefinitionEntity> {
        return searchDao.findDefinitionsLike(dictionaryId, SearchType.buildSearchCriteria(searchType, word), computeOffset(page))
    }

    private fun computeOffset(page: Int): Int {
        return maxSearchResults * (page - 1)
    }
}
